File: impacted-test-engine/src/main/java/com/teamscale/test_impacted/engine/executor/TeamscaleAgentNotifier.java
package com.teamscale.test_impacted.engine.executor;

import com.teamscale.report.testwise.model.TestExecution;
import com.teamscale.test_impacted.commons.LoggerUtils;
import com.teamscale.tia.client.ITestwiseCoverageAgentApi;
import com.teamscale.tia.client.UrlUtils;

import java.io.IOException;
import java.util.List;
import java.util.logging.Level;
import java.util.logging.Logger;

/** Communicates test start and end to the agent and the end of the overall test execution. */
public class TeamscaleAgentNotifier {

	private static final Logger LOGGER = LoggerUtils.getLogger(TeamscaleAgentNotifier.class);

	/** A list of API services to signal test start and end to the agent. */
	private final List<ITestwiseCoverageAgentApi> testwiseCoverageAgentApis;

	/**
	 * Whether only a part of the tests is being executed (<code>true</code>) or whether all tests are executed
	 * (<code>false</code>).
	 */
	private final boolean partial;

	public TeamscaleAgentNotifier(List<ITestwiseCoverageAgentApi> testwiseCoverageAgentApis, boolean partial) {
		this.testwiseCoverageAgentApis = testwiseCoverageAgentApis;
		this.partial = partial;
	}

	/** Reports the start of a test to the Teamscale JaCoCo agent. */
	public void startTest(String testUniformPath) {
		try {
			for (ITestwiseCoverageAgentApi apiService : testwiseCoverageAgentApis) {
				apiService.testStarted(UrlUtils.encodeUrl(testUniformPath)).execute();
			}
		} catch (IOException e) {
			LOGGER.log(Level.SEVERE, e, () -> "Error while calling service api.");
		}
	}

	/** Reports the end of a test to the Teamscale JaCoCo agent. */
	public void endTest(String testUniformPath, TestExecution testExecution) {
		try {
			for (ITestwiseCoverageAgentApi apiService : testwiseCoverageAgentApis) {
				if (testExecution == null) {
					apiService.testFinished(UrlUtils.encodeUrl(testUniformPath)).execute();
				} else {
					apiService.testFinished(UrlUtils.encodeUrl(testUniformPath), testExecution).execute();
				}
			}
		} catch (IOException e) {
			LOGGER.log(Level.SEVERE, e, () -> "Error contacting test wise coverage agent.");
		}
	}

	/** Reports the end of the test run to the Teamscale JaCoCo agent. */
	public void testRunEnded() {
		try {
			for (ITestwiseCoverageAgentApi apiService : testwiseCoverageAgentApis) {
				apiService.testRunFinished(partial).execute();
			}
		} catch (IOException e) {
			LOGGER.log(Level.SEVERE, e, () -> "Error contacting test wise coverage agent.");
		}
	}
}

File: impacted-test-engine/src/main/kotlin/com/teamscale/test_impacted/engine/executor/TeamscaleAgentNotifier.kt
package com.teamscale.test_impacted.engine.executor

import com.teamscale.report.testwise.model.TestExecution
import com.teamscale.test_impacted.commons.LoggerUtils.createLogger
import com.teamscale.tia.client.ITestwiseCoverageAgentApi
import com.teamscale.tia.client.UrlUtils.encodeUrl
import java.io.IOException
import java.util.logging.Level

/**
 * A notifier class responsible for communicating with the Teamscale JaCoCo agent in test-wise coverage mode.
 * It sends signals for test start, test end, and test run completion events to the specified APIs.
 *
 * @param testwiseCoverageAgentApis A list of API services used to signal test actions to the Teamscale agent.
 * @param partial Indicates whether only a subset of tests is executed (`true`) or all tests are executed (`false`).
 */
open class TeamscaleAgentNotifier(
	private val testwiseCoverageAgentApis: List<ITestwiseCoverageAgentApi>,
	private val partial: Boolean
) {
	companion object {
		private val LOG = createLogger()
	}

	/** Reports the start of a test to the Teamscale JaCoCo agent.  */
	open fun startTest(testUniformPath: String) {
		try {
			testwiseCoverageAgentApis.forEach { apiService ->
				apiService.testStarted(testUniformPath.encodeUrl()).execute()
			}
		} catch (e: IOException) {
			LOG.log(
				Level.SEVERE, e
			) { "Error while calling service api." }
		}
	}

	/** Reports the end of a test to the Teamscale JaCoCo agent.  */
	open fun endTest(testUniformPath: String, testExecution: TestExecution?) {
		try {
			testwiseCoverageAgentApis.forEach { apiService ->
				val url = testUniformPath.encodeUrl()
				if (testExecution == null) {
					apiService.testFinished(url).execute()
				} else {
					apiService.testFinished(url, testExecution).execute()
				}
			}
		} catch (e: IOException) {
			LOG.log(
				Level.SEVERE, e
			) { "Error contacting test wise coverage agent." }
		}
	}

	/** Reports the end of the test run to the Teamscale JaCoCo agent.  */
	open fun testRunEnded() {
		try {
			testwiseCoverageAgentApis.forEach { apiService ->
				apiService.testRunFinished(partial).execute()
			}
		} catch (e: IOException) {
			LOG.log(
				Level.SEVERE, e
			) { "Error contacting test wise coverage agent." }
		}
	}
}
